File: python/techai_webutils/src/techai_webutils/clients/rpc/grpc/interceptors/metrics.py
"""gRPC metrics interceptor."""

from __future__ import annotations

import time
from typing import TYPE_CHECKING

import grpc

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable

    from techai_webutils.core.interfaces.metrics import MetricCounter, MetricHistogram


class MetricsInterceptor(grpc.aio.UnaryUnaryClientInterceptor):  # type: ignore[misc]
    """Async client interceptor that records gRPC call counts and latency."""

    def __init__(self, counter: MetricCounter, histogram: MetricHistogram) -> None:
        """Store the per-call ``counter`` (by method + status) and latency ``histogram`` (by method)."""
        self._counter = counter
        self._histogram = histogram

    async def intercept_unary_unary(  # type: ignore[override]
        self,
        continuation: Callable[..., Awaitable[grpc.aio.UnaryUnaryCall]],
        client_call_details: grpc.aio.ClientCallDetails,
        request: object,
    ) -> object:
        """Intercept an async unary-unary call and record count and latency metrics."""
        method = client_call_details.method
        start = time.monotonic()
        try:
            call = await continuation(client_call_details, request)
            await call
            self._counter.increment(method=method, status="ok")
        except grpc.RpcError as e:
            self._counter.increment(method=method, status=str(e.code()))
            raise
        finally:
            self._histogram.observe(time.monotonic() - start, method=method)
        return call

File: go/clients/transport/connect/interceptors/metrics.go
package interceptors

import (
	"context"
	"time"

	"connectrpc.com/connect"
	"github.com/gt-tech-ai/knowledge-engine/go/core/interfaces"
)

// defaultDurationBuckets are the default histogram bucket boundaries for RPC
// durations (in seconds). Matches Prometheus DefBuckets.
var defaultDurationBuckets = []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10}

// metricsInterceptor records RPC metrics (request count + duration) for BOTH unary and
// server-streaming RPCs. A stream is measured as one RPC — one counter increment and one
// duration observation for the whole stream, keyed on the procedure.
//
// It is used on both the server and the client interceptor chains (ServerBuilder + the
// ClientBuilder), so WrapStreamingClient is intentionally a no-op: client-streaming
// metrics are a deliberate, currently-untracked scope boundary (only the
// server-streaming handler is covered), NOT because this interceptor is server-side.
type metricsInterceptor struct {
	// requestsTotal counts RPCs, labelled by service, method, and outcome.
	requestsTotal interfaces.Counter
	// requestDuration records RPC latency as a histogram.
	requestDuration interfaces.Histogram
}

// MetricsInterceptor creates an interceptor that records RPC metrics using the
// interfaces.Metrics abstraction. Records total requests (with service, method,
// and code labels) and request duration (with service and method labels), for unary
// and server-streaming RPCs alike.
func MetricsInterceptor(m interfaces.Metrics) connect.Interceptor {
	return metricsInterceptor{
		requestsTotal: m.Counter(
			"connect_rpc_requests_total",
			"Total number of Connect RPC requests",
			"service", "method", "code",
		),
		requestDuration: m.Histogram(
			"connect_rpc_duration_seconds",
			"Connect RPC request duration in seconds",
			defaultDurationBuckets,
			"service", "method",
		),
	}
}

// record observes the duration and increments the request counter for one completed RPC
// (unary or stream). Shared by both wrap paths.
func (i metricsInterceptor) record(procedure string, start time.Time, err error) {
	service, method := parseConnectProcedure(procedure)
	i.requestDuration.Observe(time.Since(start).Seconds(), service, method)
	code := "ok"
	if err != nil {
		code = connect.CodeOf(err).String()
	}
	i.requestsTotal.Inc(service, method, code)
}

// WrapUnary records metrics for a unary RPC.
func (i metricsInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
	return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
		start := time.Now()
		resp, err := next(ctx, req)
		i.record(req.Spec().Procedure, start, err)
		return resp, err
	}
}

// WrapStreamingClient is a no-op (see the type doc: client-streaming metrics are out of
// F6's scope).
func (i metricsInterceptor) WrapStreamingClient(
	next connect.StreamingClientFunc,
) connect.StreamingClientFunc {
	return next
}

// WrapStreamingHandler records one counter increment + one duration observation for a
// whole server-streaming RPC.
func (i metricsInterceptor) WrapStreamingHandler(
	next connect.StreamingHandlerFunc,
) connect.StreamingHandlerFunc {
	return func(ctx context.Context, conn connect.StreamingHandlerConn) error {
		start := time.Now()
		err := next(ctx, conn)
		i.record(conn.Spec().Procedure, start, err)
		return err
	}
}
